[1LP][RFR] provision cfme from azure

lina-nikiforova opened the pull request:
{{ pytest: -v cfme/tests/cloud/test_providers.py::test_create_azure_vm_from_azure_image --use-provider=complete  --long-running }}

Automation for `test_create_azure_vm_from_azure_image`. In fact, also automates `test_upload_azure_image_to_azure` as it's part of the test.

dajoRH commented:
Lint report for commit e9e422f15bdfb4b1aaf8d806afe040e166a79b36:

`cfme/tests/cloud/test_providers.py`:
- :bangbang: Line 77:90: **E226** *missing whitespace around arithmetic operator*
- :bangbang: Line 80:78: **E231** *missing whitespace after ','*
- :bangbang: Line 90:76: **E226** *missing whitespace around arithmetic operator*
- :heavy_exclamation_mark: Line 97:84: **W291** *trailing whitespace*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -36,6 +41,84 @@ def enable_regions(provider):
+@pytest.fixture(scope='module')
+def az_pwsh_vm():
+    # azure_pwsh contains powershell and necessary scripts
+    provider = get_crud_by_name('rhv42')
```
no hardcoding a provider name here, needs to be picked from cfme_data.management_systems via looking for 'powershell_vm'.

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -36,6 +41,84 @@ def enable_regions(provider):
+        return collection.instantiate(pwsh_vm, provider)
+    except IndexError:
+        from cfme.exceptions import ItemNotFound
+        raise ItemNotFound('VM with powershell not found!')
```
This should be a `pytest.skip` call, we don't have the resource necessary to run the test.

dajoRH commented:
Lint report for commit 83744d1dc143a0ad86a174e4a5f3067f51a7af18:

Lint OK! :cake: :punch: :cookie: .

* 3861a14d8e5b9dac0cb298064654eaaf005c30c1
  * Merge commit detected!!! :bangbang:

**The PR might not be acceptable in this state** :rage4:.
There is a merge commit in the branch which we do not like in PRs so please, get rid of it.

*CFME QE Bot*

digitronik commented on cfme/tests/cloud/test_providers.py:
Q: just suspect why rhv provider?

lina-nikiforova replied:
@digitronik because I need to use this provider as it has the necessary template to provision a VM with powershell and powershell scripts.

mshriver replied:
This should be written a bit more generically, as the fact that CFME QE has this template on a RHEV provider is unique to us.

I think there's an assumption here, that's not immediately apparent to me - there is an azure account required to run this.  Is the azure powershell VM already setup with the account credentials? 

I think that the fixtures for this should be structured to look for a specific definition in cfme-qe-yamls of where an azure powershell vm can be run. This could take different forms, but I think it should use the current structure of `provider.data.templates` to list the template name and creds under a specific key (following the pattern of full_template, console_template, small_template, etc).

The fixtures should look to identify any provider that has one of these VM templates defined, and somehow pick one.
The expected setup of the azure VM (credentials, installed packages) should be defined at least in a docblock, even better as a cloud-init file.

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -36,6 +46,107 @@ def enable_regions(provider):
+    # azure_pwsh contains powershell and necessary scripts
+    vm_name = random_vm_name(context="pwsh")
+    template = provider.data.get('powershell_vm')
+    vm = provider.appliance.collections.infra_vms.instantiate(vm_name, provider, template)
```
Why infra_vms collection here, for what I'm assuming is a fixture used by tests parametrized on Azure?

You can use `appliance.provider_based_collection` to get the right collection type.

As a second note, although the ProviderFilter required_fields is used to only parametrize on providers with 'powershell_vm', I think it would be good to wrap that reference in a try/except that would skip the test should this fixture be used in other test functions that forget to include this required_field.

lina-nikiforova replied:
@mshriver right, I started to mess around with that because I wanted to provision a preallocated disk, however, it appears to be not supported in RHV, so I'm reverting to my original implementation

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -36,6 +46,107 @@ def enable_regions(provider):
+    if ip_of_recourse is not None:
+        pytest.skip("The resource is taken by some other VM in Azure")
+    stream = appliance.version.stream()
+    url = conf.cfme_data['basic_info']['cfme_images_url'][stream]
```
I think this check should be wrapped for a pytest.skip call when the url match isn't found for the stream

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -36,6 +46,107 @@ def enable_regions(provider):
+        """wget -qO- {url} | grep -Po '(?<=href=")[^"]*' | grep azure""".format(url=url),
+        timeout=30).output.strip()
+    image_url = '{}/{}'.format(url, image)
+    pwsh_ssh.run_command("cd /mnt/space && wget {image}".format(image=image_url), timeout=180)
```
Consider defining the first part of this command in a variable and formatting it into the strings.

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -36,6 +46,107 @@ def enable_regions(provider):
+@pytest.fixture(scope='function')
+def upload_image_to_azure(cfme_vhd, pwsh_ssh):
+    pwsh_ssh.run_command(
+        """cd /mnt/space && sed -i "1s/.*/\$BlobNameSource = \\"{vhd}\\"/" upload_vhd.ps1"""
```
The more I see /mnt/space used, the more I think this should be a module-level 'constant', something like `CD_SPACE`

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -959,4 +1038,34 @@ def test_create_azure_vm_from_azure_image():
-    pass
+    app = appliance.IPAppliance.from_url(vm_ip)
+
+    username = credentials['azure_appliance']['username']
```
Is this a different credential than the one use for the powershell VM itself? I'm not quite clear what credential is being pulled here and used for the SSH client to that VM's IP.  Can this credential key be defined with the template name (see above comment about provider parametrization)

lina-nikiforova replied:
@mshriver yes, it's different. I'm adding a comment there.

lina-nikiforova commented:
@mshriver thanks a lot for the comments, as I moved from having one VM to the template, I actually missed connecting to the Azure account :woman_facepalming: 
I hope I addressed other concerns as well

dajoRH commented:
Lint report for commit 01d81415f612d86fd49761c8c14c61912de1cbbe:

`cfme/tests/cloud/test_providers.py`:
- :red_circle: Line 24:1: **F401** *'cfme.infrastructure.provider.rhevm.RHEVMProvider' imported but unused*
- :red_circle: Line 26:1: **F401** *'cfme.markers.env_markers.provider.ONE_PER_TYPE' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

dajoRH commented:
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

lina-nikiforova commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -22,20 +23,163 @@
+
+    if not provider_key:
+        pytest.skip("There's no provider that contains a template with powershell")
+    provider = get_crud_by_name(provider_key)
```
@izapolsk well, this works, it replaced the hardcoded provider as I did before https://github.com/ManageIQ/integration_tests/pull/8583#discussion_r266860703

mshriver replied:
This is only working by chance, when the 'name' field matches the provider key. You need to use `get_crud` here.

```
In [1]: providers.get_crud_by_name('rhv43')

Out[1]: RHEVMProvider(endpoints={'default': <cfme.infrastructure.provider.rhevm.RHEVMEndpoint object at 0x7f6664440110>, 'candu': <cfme.common.provider.CANDUEndpoint object at 0x7f6664440250>}, name='rhv43', key='rhv43', zone='default', start_ip='<REDACTED>', end_ip='<REDACTED>', provider_data=None)

In [2]: providers.get_crud_by_name('vsphere67-ims')
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-2-05d91c5e044c> in <module>()
----> 1 providers.get_crud_by_name('vsphere67-ims')

/home/mshriver/repos/integration_tests/cfme/utils/providers.py in get_crud_by_name(provider_name)
    341         if provider_data.get("name") == provider_name:
    342             return get_crud(provider_key)
--> 343     raise NameError("Could not find provider {}".format(provider_name))
    344 
    345 

NameError: Could not find provider vsphere67-ims

In [3]: providers.get_crud('vsphere67-ims')
Out[3]: VMwareProvider(endpoints={'default': <cfme.infrastructure.provider.virtualcenter.VirtualCenterEndpoint object at 0x7f6664421d90>}, name='vSphere 6.7 IMS', key='vsphere67-ims', zone='default', start_ip='<REDACTED>', end_ip='<REDACTED>', provider_data=None)
```

mshriver replied:
Please see my comment above though, as you should not have to actually call `get_crud` at all when you're just using `cfme.utils.providers` methods.

lina-nikiforova commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -22,20 +23,163 @@
+    for provider, values in providers_data.items():
+        templates = values.get('templates')
+        if templates:
+            az_template = templates.get('powershell_vm')
```
@izapolsk no, there's a template, from which we provision the VM. To recreate it we can use #8663

lina-nikiforova commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -760,9 +965,12 @@ def test_upload_azure_image_to_azure():
 
 @pytest.mark.manual
```
thanks! This is the result of my rebasing skills, my bad.

lina-nikiforova commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -22,20 +23,163 @@
+        """wget -qO- {url} | grep -Po '(?<=href=")[^"]*' | grep azure""".format(url=url),
+        timeout=30).output.strip()
+    image_url = '{}/{}'.format(url, image)
+    pwsh_ssh.run_command("{cd} wget {image}".format(cd=CD_SPACE, image=image_url), timeout=180)
```
@izapolsk  can you explain the benefits? Of course, I know the arguments but to me the current way is more clear.

mshriver replied:
I think I agree with @izapolsk here, it would be better to use such options (and in other commands provide direct paths).  There are no assertions against the ssh commands made below, and if the `cd` fails for some reason, all the subsequent commands will fail.

To me its more clear to provide full paths to the files we're manipulating, and you should be able to construct those with the constant.

lina-nikiforova replied:
@mshriver ok, I'm changing it

izapolsk replied:
@lina-nikiforova, there are several reasons why I'd better see using arguments there.
 main reason is that you rely on default behavior of tools which can be easily changed and may vary from system to system or between versions. so, you need to exactly specify what you want this tool to do.

lina-nikiforova commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -22,20 +23,163 @@
+    Connect to Azure account to run further scripts, see
+    https://docs.microsoft.com/en-us/powershell/azure/authenticate-azureps
+    """
+    pwsh_ssh.run_command("{} pwsh connect_account.ps1".format(CD_SPACE), timeout=180)
```
added

lina-nikiforova commented:
@izapolsk Thanks for the review! I'd also like to consider https://github.com/ManageIQ/integration_tests/pull/8583/files#diff-4f10bc4cf82a0f08488c8079d4cd0085R934

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -22,20 +23,164 @@
+        if templates:
+            az_template = templates.get('powershell_vm')
+            if az_template:
+                provider_key = provider
```
Can't you call `break` at this point, since you found a `provider_key` with the template defined? No need to continue iterating through the list.

lina-nikiforova replied:
definitely, I missed that

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -22,20 +23,164 @@
 
 
+@pytest.fixture(scope='function')
+def az_pwsh_vm(appliance, request):
```
request is unused here, are you expecting indirect parametrization?

lina-nikiforova replied:
no, I wanted to add finalizer but implemented it using `yield`, so it's indeed not needed

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -22,20 +23,164 @@
+            if az_template:
+                provider_key = provider
+
+    if not provider_key:
```
`provider_key` is set within the `if` blocks above, and if there was no matching provider found, this will trigger an `UnboundLocalError`.

I think you need a bit more efficient structure here to handle these cases. Take advantage of the tools in `cfme.utils.providers`.

```
filter = cfme.utils.providers.ProviderFilter(required_fields=[['templates', 'powershell_vm']])
provider_objects_list = cfme.utils.providers.list_providers(filters=[filter])
```

lina-nikiforova replied:
@mshriver huge thanks! this is so much better.

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -727,42 +871,74 @@ def test_domain_id_validation(request, provider):
     """
-    Upload image copied from RHCF3-11271 to azure using powershell
-    similiar the script in the setup section.
+    https://bugzilla.redhat.com/show_bug.cgi?id=1520196
```
Please use `@pytest.mark.meta(automates=[BZ(1520196)])` to indicate test coverage.

lina-nikiforova replied:
hmmm, as it's a manual test, I don't think it's correct.
I'm adding it like this
```
    Bugzilla:
        1520196
```
as I see it in other _manual_ tests as well

mshriver commented on cfme/tests/cloud/test_providers.py:
Need `expectedResults` for a manual test case, so its clear what is being validated or checked here.

mshriver commented on cfme/tests/cloud/test_providers.py:
Please use docblock section to list BZ's associated with the manual case (no pytest marker at this time for manual cases)

```
Bugzilla:
    1357086
```

mshriver commented on cfme/tests/cloud/test_providers.py:
testSteps and expectedResults please, even if they're short.

lina-nikiforova replied:
well, I didn't want to do changes to manual tests as I think it's out of scope of this PR, but why not

mshriver commented on cfme/tests/cloud/test_providers.py:
Please split this to an expectedResults block, with a bit more detail on what is expected to be visible/not visible. Should enabling market place de-list the private images? Not sure from this test case.

mshriver commented on cfme/tests/cloud/test_providers.py:
Please use `Bugzilla` docblock section.

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -831,4 +1007,37 @@ def test_create_azure_vm_from_azure_image():
         title: Create Azure VM from Azure image
     """
-    pass
+    app = appliance.IPAppliance.from_url(vm_ip)
```
Consider using `appliance.Appliance.from_provider()` here, though you may not want to given your fixtures for `vm_ip`

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -831,4 +1007,37 @@ def test_create_azure_vm_from_azure_image():
+    username = credentials['azure_appliance']['username']
+    password = credentials['azure_appliance']['password']
+
+    with ssh.SSHClient(hostname=vm_ip,
```
I would love to see an enhancement that allows for passing username/password to `IPAppliance.__init__`, setting local vars during class construction.  Some `IPAppliance.username` and `IPAppliance.password` properties could return these local vars (`self._username`).

Then `IPAppliance.ssh_client` could use those properties, defaulting to the current conf file lookup when nothing is passed.

Then, when you create an appliance object here to test against, you can still use `app.ssh_client`, and it will automatically use the passed credentials instead of the default.

Just a thought, for a framework enhancement that would assist with this kind of testing.  There's already some code there for handling openshift creds for appliances.

lina-nikiforova replied:
@mshriver totally agree, Mike. Although I would prefer much more to have smaller PRs and this one is growing already to editing manual tests that I didn't touch here. I can try to raise a follow-up PR to do such changes.

mshriver replied:
oh, absolutely don't add that here, was just throwing the idea out there :)

mshriver commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -831,4 +1007,37 @@ def test_create_azure_vm_from_azure_image():
+
+        # restart sshd to apply configuration changes
+        restart = app_ssh_client.run_command(
+            'echo {} | sudo -S systemctl restart sshd'.format(password), ensure_user=True)
```
Consider adding a `SystemdService` attribute to IPAppliance for `sshd`

izapolsk commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -22,20 +23,160 @@
+    image = pwsh_ssh.run_command(
+        """wget -qO- {url} | grep -Po '(?<=href=")[^"]*' | grep azure""".format(url=url),
+        timeout=30).output.strip()
+    image_url = '{}/{}'.format(url, image)
```
I'd consider using urljoin or something similar

lina-nikiforova replied:
@izapolsk right, makes sense

lina-nikiforova commented on cfme/tests/cloud/test_providers.py:
```diff
@@ -22,20 +23,160 @@
+    Connect to Azure account to run further scripts, see
+    https://docs.microsoft.com/en-us/powershell/azure/authenticate-azureps
+    """
+    connect = pwsh_ssh.run_command("pwsh {}connect_account.ps1".format(SPACE), timeout=180)
```
@izapolsk thanks, gonna update

lina-nikiforova commented:
@izapolsk I just pushed the change with the path as per your comments.
The resource is taken by another VM, I can look into it more later but it definitely worked when I tested.
